src/Engine/ProtoCore/DSASM/InstructionSet.cs

```diff
@@ -693,25 +937,35 @@ public static List<StackValue> BuildInvalidRegisters()
         /// type of StackValue should be AddressType.ArrayKey. 
         /// </summary>
         /// <param name="array"></param>
-        /// <param name="index"></param>
+        /// <param name="key"></param>
         /// <returns></returns>
-        public bool TryGetArrayKey(out StackValue array, out int index)
+        public bool TryGetArrayKey(out StackValue array, out int key)
         {
             array = StackValue.Null;
-            index = Constants.kInvalidIndex;
+            key = Constants.kInvalidIndex;
 
-            if (!this.IsArrayKey || opdata == Constants.kInvalidIndex)
-            {
+            if (!IsArrayKey)
                 return false;
-            }
-
+            
+            if (opdata == Constants.kInvalidIndex)
+               return false;
+            
+            // Array key information is encoded in 64bits opdata. 
+            // High 32 bits: array pointer
+            // Low 32 bits : array key
+            //
+            // TODO: find out a cleaner way to represent array key instead of
+            // using this kind of hacking.
+            var rawArrayPointer = ((ulong)opdata >> 32);
```

Comment:
As you said, this _hacking way_ is being used at two places, if we have static methods on `StackValue` that deals with conversions, wouldn't that be safer (and more testable)?